File: picasso-sample/src/main/java/com/example/picasso/SampleListDetailActivity.java
package com.example.picasso;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;
import com.squareup.picasso3.provider.PicassoProvider;

public class SampleListDetailActivity extends PicassoSampleActivity {
  @Override protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    if (savedInstanceState == null) {
      getSupportFragmentManager().beginTransaction()
          .add(R.id.sample_content, ListFragment.newInstance())
          .commit();
    }
  }

  void showDetails(String url) {
    getSupportFragmentManager().beginTransaction()
        .replace(R.id.sample_content, DetailFragment.newInstance(url))
        .addToBackStack(null)
        .commit();
  }

  public static class ListFragment extends Fragment {
    public static ListFragment newInstance() {
      return new ListFragment();
    }

    @Override public View onCreateView(LayoutInflater inflater, ViewGroup container,
        Bundle savedInstanceState) {
      final SampleListDetailActivity activity = (SampleListDetailActivity) getActivity();
      final SampleListDetailAdapter adapter = new SampleListDetailAdapter(activity);

      ListView listView = (ListView) LayoutInflater.from(activity)
          .inflate(R.layout.sample_list_detail_list, container, false);
      listView.setAdapter(adapter);
      listView.setOnScrollListener(new SampleScrollListener(activity));
      listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
        @Override
        public void onItemClick(AdapterView<?> adapterView, View view, int position, long id) {
          String url = adapter.getItem(position);
          activity.showDetails(url);
        }
      });
      return listView;
    }
  }

  public static class DetailFragment extends Fragment {
    private static final String KEY_URL = "picasso:url";

    public static DetailFragment newInstance(String url) {
      Bundle arguments = new Bundle();
      arguments.putString(KEY_URL, url);

      DetailFragment fragment = new DetailFragment();
      fragment.setArguments(arguments);
      return fragment;
    }

    @Override public View onCreateView(LayoutInflater inflater, ViewGroup container,
        Bundle savedInstanceState) {
      SampleListDetailActivity activity = (SampleListDetailActivity) getActivity();

      View view = LayoutInflater.from(activity)
          .inflate(R.layout.sample_list_detail_detail, container, false);

      TextView urlView = view.findViewById(R.id.url);
      ImageView imageView = view.findViewById(R.id.photo);

      Bundle arguments = getArguments();
      String url = arguments.getString(KEY_URL);

      urlView.setText(url);
      PicassoProvider.get()
          .load(url)
          .fit()
          .tag(activity)
          .into(imageView);

      return view;
    }
  }
}

File: picasso-sample/src/main/java/com/example/picasso/SampleListDetailActivity.kt
/*
 * Copyright (C) 2022 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.example.picasso

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView.OnItemClickListener
import android.widget.ImageView
import android.widget.ListView
import android.widget.TextView
import androidx.fragment.app.Fragment

class SampleListDetailActivity : PicassoSampleActivity() {
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    if (savedInstanceState == null) {
      supportFragmentManager
        .beginTransaction()
        .add(R.id.sample_content, ListFragment.newInstance())
        .commit()
    }
  }

  fun showDetails(url: String) {
    supportFragmentManager
      .beginTransaction()
      .replace(R.id.sample_content, DetailFragment.newInstance(url))
      .addToBackStack(null)
      .commit()
  }

  class ListFragment : Fragment() {
    override fun onCreateView(
      inflater: LayoutInflater,
      container: ViewGroup?,
      savedInstanceState: Bundle?
    ): View? {
      val activity = activity as SampleListDetailActivity
      val adapter = SampleListDetailAdapter(activity)
      val listView = LayoutInflater.from(activity)
        .inflate(R.layout.sample_list_detail_list, container, false) as ListView

      listView.adapter = adapter
      listView.setOnScrollListener(SampleScrollListener(activity))
      listView.onItemClickListener = OnItemClickListener { _, _, position, _ ->
        val url = adapter.getItem(position)
        activity.showDetails(url)
      }

      return listView
    }

    companion object {
      fun newInstance(): ListFragment {
        return ListFragment()
      }
    }
  }

  class DetailFragment : Fragment() {
    override fun onCreateView(
      inflater: LayoutInflater,
      container: ViewGroup?,
      savedInstanceState: Bundle?
    ): View? {
      val activity = activity as SampleListDetailActivity
      val view = LayoutInflater.from(activity)
        .inflate(R.layout.sample_list_detail_detail, container, false)

      val urlView = view.findViewById<TextView>(R.id.url)
      val imageView = view.findViewById<ImageView>(R.id.photo)

      val url = requireArguments().getString(KEY_URL)
      urlView.text = url
      PicassoInitializer.get()
        .load(url)
        .fit()
        .tag(activity)
        .into(imageView)

      return view
    }

    companion object {
      private const val KEY_URL = "picasso:url"

      fun newInstance(url: String): DetailFragment {
        return DetailFragment().apply {
          arguments = Bundle().apply { putString(KEY_URL, url) }
        }
      }
    }
  }
}
